doc/zk-integration.md

```diff
@@ -0,0 +1,261 @@
+contract RollupContract1 is RollupContract {
+
+    function updateRollupState(
+        newVMState,
+        commTxsRollup,
+        snarkProof){
+
+        // Escape hatch is activated, switch to default sequencing mode
+        if (isEscapeHatchActivated()){
+            this.updateStateBackupSequencingMode(commTxsRollup);
+        } else { // No escape hatch, use the state of Espresso consensus
+            lightClientState = lcContract.getFinalizedState();
+            newEspressoState = lightClientState.blockCommRoot;
+            blockNumberEspresso = lightClientState.blockHeight;
+            this.updateStateFromEspresso(newEspressoState, blockNumberEspresso, commTxsRollup, newVMState, snarkProof);
+        }
+    }
+}
+```
+
+## Integration 2: Verify Espresso consensus inside the rollup circuit
+
+![image](zk-rollup-circuit.svg) **Figure 3:** Rollup circuit with additional gadgets _Espresso Consensus_, _Collect &
+Filter_, and _COMMs Equivalence_. Private inputs of the circuit are written in bold font.
+
+The circuit depicted in Figure 3 operates as follows:
+
+- The _Espresso Consensus_ gadget checks that the block commitment for Espresso block `BLOCK_NUMBER` is `blk_cm`. To
+  achieve this goal, it is required to obtain the state of the stake table from the previous HotShot epoch. Assuming the
+  rollup state is updated at least once per epoch, the commitment `blk_cm_old` will be computed from such state. Hence,
+  the private input `STAKE_TABLE_ENTRIES`, containing the list of public keys with their respective stake, can be linked
+  to the commitment `blk_cm_old` via the private inputs `STAKE_TABLE_OPENINGS`. Finally, note that the first
+  `blk_cm_old` value needs to be read from the light client contract. Afterward, no dependency on the light client
+  contract is needed.
+- The other gadgets are the same as in Integration 1.
+
+The pseudocode of the rollup contract below shows that in the case we do not rely on the Espresso light client contract
+to fetch the Espresso state, the function `updateRollupState` requires additional inputs (compared to Integration 1)
+which are `newEspressoState` and `blockNumberEspresso`.
+
+```solidity
+
+/// Does not use the Espresso Light client contract for fetching the Espresso state
+contract RollupContract2 is RollupContract {
+
+  function updateRollupState(
+    newEspressoState,
+    blockNumberEspresso,
+    newVMState,
+    commTxsRollup,
+    snarkProof){
+
+    // Escape hatch is activated, switch to default sequencing mode
+    if (isEscapeHatchActivated()){
+      this.updateStateBackupSequencingMode(commTxsRollup);
+    } else { // No escape hatch, use the state of Espresso consensus
+      this.updateStateFromEspresso(newEspressoState, blockNumberEspresso, commTxsRollup, newVMState, snarkProof);
+    }
+  }
+}
+```
+
+[^1]:
+    Note that the rollup state is updated at a much lower frequency (in the order of minutes / hours) than the Espresso
```

Comment:
nit(comment): IIUC this is two different (incomparable) kinds of "state updates".

Reply:
What do you mean by incomparable? Do you suggest updating the footnote in some way?